File: Java/src/main/java/codingdojo/XMLExporter.java
package codingdojo;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Collection;
import java.util.Date;

public class XMLExporter {

    public static String exportFull(Collection<Order> orders) {
        StringBuilder xml = new StringBuilder();
        xml.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
        xml.append("<orders>");
        for (Order order : orders) {
            xml.append("<order");
            xml.append(" id='");
            xml.append(order.getId());
            xml.append("'>");
            for (Product product : order.getProducts()) {
                xml.append("<product");
                xml.append(" id='");
                xml.append(product.getId());
                xml.append("'");
                if (product.isEvent()) {
                    xml.append(" stylist='");
                    xml.append(stylistFor(product));
                    xml.append("'");
                }

                if (product.getWeight() > 0) {
                    xml.append(" weight='");
                    xml.append(product.getWeight());
                    xml.append("'");
                }

                xml.append(">");
                xml.append("<price");
                xml.append(" currency='");
                xml.append(product.getPrice().getCurrency());
                xml.append("'>");
                xml.append(product.getPrice().getAmount());
                xml.append("</price>");
                xml.append(product.getName());
                xml.append("</product>");
            }

            xml.append("</order>");
        }

        xml.append("</orders>");
        return xml.toString();
    }

    public static String exportTaxDetails(Collection<Order> orders) {
        NumberFormat formatter = new DecimalFormat("#0.00");
        StringBuilder xml = new StringBuilder();
        xml.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
        xml.append("<orderTax>");
        for (Order order : orders) {
            xml.append("<order");
            xml.append(" date='");
            xml.append(Util.toIsoDate(order.getDate()));
            xml.append("'");
            xml.append(">");
            double tax = 0D;
            for (Product product : order.getProducts()) {
                xml.append("<product");
                xml.append(" id='");
                xml.append(product.getId());
                xml.append("'");
                xml.append(">");
                xml.append(product.getName());
                xml.append("</product>");
                if (product.isEvent())
                    tax += product.getPrice().getAmountInCurrency("USD") * 0.25;
                else
                    tax += product.getPrice().getAmountInCurrency("USD") * 0.175;

            }

            xml.append("<orderTax currency='USD'>");
            if (order.getDate().before(Util.fromIsoDate("2018-01-01T00:00Z")))
                tax += 10;
            else
                tax += 20;
            xml.append(formatter.format(tax));
            xml.append("</orderTax>");
            xml.append("</order>");
        }
        
        double totalTax = TaxCalculator.calculateAddedTax(orders);
        xml.append(formatter.format(totalTax));
        xml.append("</orderTax>");
        return xml.toString();
    }

    public static String exportStore(Store store) {
        StringBuffer xml = new StringBuffer();
        xml.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");

        xml.append("<store");
        xml.append(" name='");
        xml.append(store.getName());
        xml.append("'");
        xml.append(">");
        for (Product product : store.getStock()) {
            xml.append("<product");
            xml.append(" id='");
            xml.append(product.getId());
            xml.append("'");
            if (product.isEvent()) {
                xml.append(" location='");
                xml.append(store.getName());
                xml.append("'");
            } else {
                xml.append(" weight='");
                xml.append(product.getWeight());
                xml.append("'");
            }

            xml.append(">");
            xml.append("<price");
            xml.append(" currency='");
            xml.append(product.getPrice().getCurrency());
            xml.append("'>");
            xml.append(product.getPrice().getAmount());
            xml.append("</price>");
            xml.append(product.getName());
            xml.append("</product>");
        }

        xml.append("</store>");

        return xml.toString();
    }

    public static String exportHistory(Collection<Order> orders) {
        StringBuilder xml = new StringBuilder();
        xml.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
        xml.append("<orderHistory");
        xml.append(" createdAt='");
        Date now = new Date();
        xml.append(Util.toIsoDate(now));
        xml.append("'");
        xml.append(">");
        for (Order order : orders) {
            xml.append("<order");
            xml.append(" date='");
            xml.append(Util.toIsoDate(order.getDate()));
            xml.append("'");
            xml.append(" totalDollars='");
            xml.append(order.totalDollars());
            xml.append("'>");
            for (Product product : order.getProducts()) {
                xml.append("<product");
                xml.append(" id='");
                xml.append(product.getId());
                xml.append("'");
                xml.append(">");
                xml.append(product.getName());
                xml.append("</product>");
            }
            
            xml.append("</order>");
        }
        
        xml.append("</orderHistory>");
        return xml.toString();
    }

    private static String stylistFor(Product product) {
        return "Celeste Pulchritudo"; // in future we will look up the name of the stylist from the database
    }

}

File: TypeScript/src/XMLExporter.ts
import { Order } from "./Order";
import { Product } from "./Product";
import { Store } from "./Store";
import { TaxCalculator } from "./TaxCalculator";
import { Util } from "./Util";

export class XMLExporter {

    static exportFull(orders: Order[]): string {
        let xml: string[] = [];
        xml.push(`<?xml version="1.0" encoding="UTF-8"?>`);
        xml.push("<orders>");

        for (const order of orders) {
            xml.push(`<order id='${order.getId()}'>`);

            for (const product of order.getProducts()) {
                xml.push(`<product id='${product.getId()}'`);

                if (product.isEvent()) {
                    xml.push(` stylist='${XMLExporter.stylistFor(product)}'`);
                }

                if (product.getWeight() > 0) {
                    xml.push(` weight='${product.getWeight()}'`);
                }

                xml.push(">");
                xml.push(`<price currency='${product.getPrice().getCurrency()}'>${product.getPrice().getAmount()}</price>`);
                xml.push(product.getName());
                xml.push("</product>");
            }

            xml.push("</order>");
        }

        xml.push("</orders>");
        return xml.join("\n");
    }

    static exportTaxDetails(orders: Order[]): string {
        function formatNumber(value: number): string {
            return value.toFixed(2);
        }

        let xml: string[] = [];
        xml.push(`<?xml version="1.0" encoding="UTF-8"?>`);
        xml.push("<orderTax>");

        for (const order of orders) {
            xml.push(`<order date='${Util.toIsoDate(order.getDate())}'>`);
            let tax = 0;

            for (const product of order.getProducts()) {
                xml.push(`<product id='${product.getId()}'>${product.getName()}</product>`);

                if (product.isEvent()) {
                    tax += product.getPrice().getAmountInCurrency("USD") * 0.25;
                } else {
                    tax += product.getPrice().getAmountInCurrency("USD") * 0.175;
                }
            }

            if (order.getDate() < Util.fromIsoDate("2018-01-01T00:00Z")) {
                tax += 10;
            } else {
                tax += 20;
            }

            xml.push(`<orderTax currency='USD'>${formatNumber(tax)}</orderTax>`);
            xml.push("</order>");
        }

        const totalTax = TaxCalculator.calculateAddedTax(orders);
        xml.push(formatNumber(totalTax));
        xml.push("</orderTax>");

        return xml.join("\n");
    }

    static exportStore(store: Store): string {
        let xml: string[] = [];
        xml.push(`<?xml version="1.0" encoding="UTF-8"?>`);
        xml.push(`<store name='${store.getName()}'>`);

        for (const product of store.getStock()) {
            xml.push(`<product id='${product.getId()}'`);

            if (product.isEvent()) {
                xml.push(` location='${store.getName()}'`);
            } else {
                xml.push(` weight='${product.getWeight()}'`);
            }

            xml.push(">");
            xml.push(`<price currency='${product.getPrice().getCurrency()}'>${product.getPrice().getAmount()}</price>`);
            xml.push(product.getName());
            xml.push("</product>");
        }

        xml.push("</store>");
        return xml.join("\n");
    }

    static exportHistory(orders: Order[]): string {
        let xml: string[] = [];
        xml.push(`<?xml version="1.0" encoding="UTF-8"?>`);
        xml.push(`<orderHistory createdAt='${Util.toIsoDate(new Date())}'>`);

        for (const order of orders) {
            xml.push(`<order date='${Util.toIsoDate(order.getDate())}' totalDollars='${order.totalDollars()}'>`);

            for (const product of order.getProducts()) {
                xml.push(`<product id='${product.getId()}'>${product.getName()}</product>`);
            }

            xml.push("</order>");
        }

        xml.push("</orderHistory>");
        return xml.join("\n");
    }


    private static stylistFor(product: Product): string {
        return "Celeste Pulchritudo"; // In the future, look up the name of the stylist from the database
    }

}
